File: linux/overlay.py
import time
from dataclasses import dataclass, field
from uuid import uuid4

from PyQt6.QtCore import Qt, QTimer
from PyQt6.QtGui import QColor, QFont, QFontMetrics, QPainter, QPainterPath, QPen, QGuiApplication
from PyQt6.QtWidgets import QWidget

from settings import settings


@dataclass
class KeyDisplay:
    uid: str = field(default_factory=lambda: uuid4().hex)
    text: str = ""
    expire: float = 0.0


class KeyBubble(QWidget):
    def __init__(self, text: str, parent=None):
        super().__init__(parent)
        self.text = text
        self.setAttribute(Qt.WidgetAttribute.WA_TransparentForMouseEvents)
        self._update_size()

    def _update_size(self):
        font = QFont("sans-serif", settings["font_size"])
        font.setBold(True)
        fm = QFontMetrics(font)
        self.setFixedSize(fm.horizontalAdvance(self.text) + 40, fm.height() + 24)

    def paintEvent(self, _event):
        p = QPainter(self)
        p.setRenderHint(QPainter.RenderHint.Antialiasing)

        w, h = float(self.width()), float(self.height())
        radius = float(settings["corner_radius"])

        bg = QColor(settings["bg_color"])
        bg.setAlphaF(settings["bg_opacity"])
        path = QPainterPath()
        path.addRoundedRect(0.0, 0.0, w, h, radius, radius)
        p.fillPath(path, bg)

        bw = settings["border_width"]
        if bw > 0:
            bc = QColor(settings["border_color"])
            bc.setAlphaF(settings["border_opacity"])
            p.setPen(QPen(bc, bw))
            p.drawRoundedRect(int(bw / 2), int(bw / 2), int(w - bw), int(h - bw), radius, radius)

        tc = QColor(settings["text_color"])
        tc.setAlphaF(settings["text_opacity"])
        p.setPen(tc)
        font = QFont("sans-serif", settings["font_size"])
        font.setBold(True)
        p.setFont(font)
        p.drawText(self.rect(), Qt.AlignmentFlag.AlignCenter, self.text)
        p.end()


class OverlayWindow(QWidget):
    SPACING = 8
    MARGIN = 50

    def __init__(self):
        super().__init__()
        self.setWindowFlags(
            Qt.WindowType.FramelessWindowHint
            | Qt.WindowType.WindowStaysOnTopHint
            | Qt.WindowType.Tool
            | Qt.WindowType.WindowTransparentForInput
            | Qt.WindowType.WindowDoesNotAcceptFocus
        )
        self.setAttribute(Qt.WidgetAttribute.WA_TranslucentBackground)
        self.setAttribute(Qt.WidgetAttribute.WA_TransparentForMouseEvents)
        self.setAttribute(Qt.WidgetAttribute.WA_ShowWithoutActivating)

        self._keys: list[tuple[KeyDisplay, KeyBubble]] = []

        self._timer = QTimer(self)
        self._timer.timeout.connect(self._cleanup)
        self._timer.start(100)

        settings.changed.connect(self._reposition)
        self._reposition()

    def _reposition(self):
        screen = QGuiApplication.primaryScreen()
        if screen:
            self.setGeometry(screen.availableGeometry())
            self._relayout()

    def add_key(self, text: str):
        kd = KeyDisplay(text=text, expire=time.time() + settings["display_duration"])
        bubble = KeyBubble(text, self)
        bubble.show()
        self._keys.append((kd, bubble))

        while len(self._keys) > settings["max_keys"]:
            _, old = self._keys.pop(0)
            old.deleteLater()

        self._relayout()

    def _relayout(self):
        bubbles = [b for _, b in self._keys]
        if not bubbles:
            return

        pos = settings["position"]
        geo = self.geometry()
        total_h = sum(b.height() for b in bubbles) + self.SPACING * (len(bubbles) - 1)

        y = (
            self.MARGIN if "top" in pos
            else geo.height() - self.MARGIN - total_h if "bottom" in pos
            else (geo.height() - total_h) // 2
        )

        for _, bubble in self._keys:
            x = (
                self.MARGIN if "left" in pos
                else geo.width() - self.MARGIN - bubble.width() if "right" in pos
                else (geo.width() - bubble.width()) // 2
            )
            bubble.move(x, y)
            y += bubble.height() + self.SPACING

    def _cleanup(self):
        now = time.time()
        expired = [(kd, b) for kd, b in self._keys if kd.expire < now]
        if not expired:
            return
        for _, b in expired:
            b.deleteLater()
        self._keys = [(kd, b) for kd, b in self._keys if kd.expire >= now]
        self._relayout()

File: linux/src/overlay.cpp
#include "overlay.h"
#include "settings.h"

#include <QPainter>
#include <QPainterPath>
#include <QGuiApplication>
#include <QScreen>
#include <QDateTime>

static constexpr int MARGIN = 50;
static constexpr int SPACING = 8;
static constexpr int PAD_H = 20;
static constexpr int PAD_V = 12;

OverlayWindow::OverlayWindow(QWidget* parent) : QWidget(parent) {
    setWindowFlags(
        Qt::FramelessWindowHint
        | Qt::WindowStaysOnTopHint
        | Qt::X11BypassWindowManagerHint
        | Qt::WindowTransparentForInput
        | Qt::WindowDoesNotAcceptFocus
    );
    setAttribute(Qt::WA_TranslucentBackground);
    setAttribute(Qt::WA_TransparentForMouseEvents);
    setAttribute(Qt::WA_ShowWithoutActivating);

    connect(&m_timer, &QTimer::timeout, this, &OverlayWindow::cleanup);
    m_timer.start(100);

    connect(&Settings::instance(), &Settings::changed, this, &OverlayWindow::reposition);
    reposition();
}

void OverlayWindow::reposition() {
    if (auto* screen = QGuiApplication::primaryScreen())
        setGeometry(screen->availableGeometry());
    update();
}

void OverlayWindow::addKey(const QString& text) {
    auto& s = Settings::instance();
    auto expireMs = QDateTime::currentMSecsSinceEpoch() + static_cast<qint64>(s.displayDuration() * 1000);
    m_keys.push_back({text, expireMs});

    while (static_cast<int>(m_keys.size()) > s.maxKeys())
        m_keys.erase(m_keys.begin());

    update();
}

void OverlayWindow::cleanup() {
    auto now = QDateTime::currentMSecsSinceEpoch();
    auto before = m_keys.size();

    std::erase_if(m_keys, [now](const KeyDisplay& k) { return k.expireMs < now; });

    if (m_keys.size() != before)
        update();
}

std::vector<OverlayWindow::BubbleRect> OverlayWindow::layoutBubbles() {
    auto& s = Settings::instance();
    auto pos = s.position();

    QFont font("sans-serif", s.fontSize());
    font.setBold(true);
    QFontMetrics fm(font);

    std::vector<BubbleRect> bubbles;
    int totalH = 0;

    for (auto& key : m_keys) {
        int w = fm.horizontalAdvance(key.text) + PAD_H * 2;
        int h = fm.height() + PAD_V * 2;
        bubbles.push_back({{0, 0, (double)w, (double)h}, key.text});
        totalH += h;
    }
    if (!bubbles.empty())
        totalH += SPACING * (static_cast<int>(bubbles.size()) - 1);

    int yStart;
    if (pos.contains("top"))         yStart = MARGIN;
    else if (pos.contains("bottom")) yStart = height() - MARGIN - totalH;
    else                             yStart = (height() - totalH) / 2;

    int y = yStart;
    for (auto& b : bubbles) {
        int bw = static_cast<int>(b.rect.width());
        int x;
        if (pos.contains("left"))        x = MARGIN;
        else if (pos.contains("right"))  x = width() - MARGIN - bw;
        else                             x = (width() - bw) / 2;

        b.rect.moveTopLeft({(double)x, (double)y});
        y += static_cast<int>(b.rect.height()) + SPACING;
    }

    return bubbles;
}

void OverlayWindow::paintEvent(QPaintEvent*) {
    if (m_keys.empty()) return;

    auto& s = Settings::instance();
    auto bubbles = layoutBubbles();

    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing);

    QFont font("sans-serif", s.fontSize());
    font.setBold(true);
    p.setFont(font);

    double radius = s.cornerRadius();

    QColor bg(s.bgColor());
    bg.setAlphaF(s.bgOpacity());

    QColor tc(s.textColor());
    tc.setAlphaF(s.textOpacity());

    QColor bc(s.borderColor());
    bc.setAlphaF(s.borderOpacity());

    for (auto& b : bubbles) {
        QPainterPath path;
        path.addRoundedRect(b.rect, radius, radius);
        p.fillPath(path, bg);

        if (s.borderWidth() > 0) {
            p.setPen(QPen(bc, s.borderWidth()));
            p.drawRoundedRect(b.rect, radius, radius);
        }

        p.setPen(tc);
        p.drawText(b.rect, Qt::AlignCenter, b.text);
    }
}
